core/src/main/java/com/nisovin/magicspells/util/data/DataEntity.java

```diff
@@ -41,6 +43,11 @@ public class DataEntity {
 			EntityDamageEvent event = entity.getLastDamageCause();
 			return event == null ? "" : event.getDamage() + "";
 		});
+		dataElements.put("lastdamagecause.attacker", entity -> {
+			if (!(entity.getLastDamageCause() instanceof EntityDamageByEntityEvent event)) return "";
+			if (event.getDamager().customName() != null) return Util.getStrictStringFromComponent(event.getDamager().name());
+			return MagicSpells.getTargetName(event.getDamager());
```

Comment:
Currently, `MagicSpells#getTargetName` returns a configured name for an entity from `general.yml`. If not found it returns its `name` which is its display name if present or its vanilla entity name. I think this should be changed so that the target's name first returned is its display name, then its MS-configured name, then its `name`. But let's see if other devs agree.

In that case that method would go like this:
```java
public static String getTargetName(Entity target) {
	if (target instanceof Player) return target.getName();

	if (entity.customName() != null) return Util.getStrictStringFromComponent(entity.customName());
	. . .
	(entityNames)
	. . .
}
```
And the data element could directly return the call of this method.

Reply:
Returning player's custom name may be an issue for spells using %t as a parameter. For instance, if an external spell has a command that's like "tp %t 0 0 0" it may not register that player for the command. This is my thoughts on this change.

Reply:
It was changed to return player usernames exactly for that command support. That's not a change I'm proposing. I'm wondering why we shouldn't prioritise the display name over the vanilla name of non-player entities, as you wrote the data element code with this priority yourself. My snippet moved that line to the `getTargetName` method.